File: front/src/lib/import-rules/engine.ts
import type {
  Classification,
  ImportRule,
  PendingMovement,
  PersonaValue,
  RuleActions,
  RuleCondition,
  TablaDestino,
} from "./types";

function applyBaseline(movement: PendingMovement): Classification {
  return {
    tablaDestino: null,
    ignorar: false,
    destino: null,
    origen: null,
    notas: null,
    categoria: null,
    tipo: null,
    nombre: null,
    entidad: null,
    persona: movement.persona,
    importe: movement.importe,
  };
}

function extractDestinoFromContains(concepto: string, needle: string): string {
  const idx = concepto.toLowerCase().indexOf(needle.toLowerCase());
  if (idx === -1) return needle.trim();
  return concepto.slice(idx, idx + needle.length).trim();
}

function conceptoExactoValues(exact: string | string[]): string[] {
  return Array.isArray(exact) ? exact : [exact];
}

function matchesConceptoExacto(concepto: string, exact: string | string[]): boolean {
  const normalized = normalizeConcepto(concepto);
  return conceptoExactoValues(exact).some((value) => normalized === normalizeConcepto(value));
}

function conceptoContieneValues(contains: string | string[]): string[] {
  return Array.isArray(contains) ? contains : [contains];
}

function matchesConceptoContiene(concepto: string, contains: string | string[]): boolean {
  const lower = concepto.toLowerCase();
  return conceptoContieneValues(contains).some((needle) => lower.includes(needle.toLowerCase()));
}

function inferDestinoFromCondition(
  concepto: string,
  condition: RuleCondition,
  actions: RuleActions,
): string | undefined {
  if (actions.destino !== undefined) return actions.destino.trim() || undefined;
  const contains = condition.concepto_contiene;
  if (contains) {
    const matched = conceptoContieneValues(contains).find((needle) =>
      concepto.toLowerCase().includes(needle.toLowerCase()),
    );
    if (matched) return extractDestinoFromContains(concepto, matched);
  }
  const exact = condition.concepto_exacto;
  if (exact) {
    const matched = conceptoExactoValues(exact).find(
      (value) => normalizeConcepto(concepto) === normalizeConcepto(value),
    );
    if (matched) return extractDestinoFromContains(concepto, matched);
  }
  return undefined;
}

export function normalizeConcepto(value: string): string {
  return value.trim().toLowerCase();
}

/** Normaliza condiciones leídas de NocoDB (p. ej. concepto_contiene serializado como string). */
export function normalizeRuleCondition(condition: RuleCondition): RuleCondition {
  const out: RuleCondition = { ...condition };
  const contains = out.concepto_contiene;
  if (typeof contains === "string" && contains.trim().startsWith("[")) {
    try {
      const parsed: unknown = JSON.parse(contains);
      if (Array.isArray(parsed)) {
        out.concepto_contiene = parsed.map((value) => String(value));
      }
    } catch {
      /* mantener string */
    }
  }
  return out;
}

function matchConceptoRegex(concepto: string, pattern: string): RegExpMatchArray | null {
  try {
    return concepto.match(new RegExp(pattern, "i"));
  } catch {
    return null;
  }
}

function extractNombreFromRegex(concepto: string, condition: RuleCondition): string | null {
  const pattern = condition.concepto_regex;
  if (!pattern) return null;
  const match = matchConceptoRegex(concepto, pattern);
  if (!match) return null;
  const group = condition.nombre_grupo ?? 1;
  const value = match[group];
  return value?.trim() || null;
}

export function applyRules(movement: PendingMovement, rules: ImportRule[]): Classification {
  return applyRulesWithMeta(movement, rules).classification;
}

export interface RuleMatchResult {
  classification: Classification;
  reglaId: string | null;
  reglaNombre: string | null;
  reglaPrioridad: number;
}

/** Clasifica un movimiento y devuelve la regla de mayor prioridad que coincidió. */
export function applyRulesWithMeta(
  movement: PendingMovement,
  rules: ImportRule[],
): RuleMatchResult {
  let classified = applyBaseline(movement);
  const accountId = movement.metadata.account_id ?? "";
  let lastMatched: ImportRule | null = null;

  const applicable = rules
    .filter((rule) => rule.active && ruleMatches(rule, movement, accountId))
    .sort((a, b) => a.priority - b.priority);

  for (const rule of applicable) {
    let actions = rule.actions;
    const extractedNombre = extractNombreFromRegex(movement.concepto, rule.condition);
    if (extractedNombre !== null && actions.nombre === undefined) {
      actions = { ...actions, nombre: extractedNombre };
    }
    const inferredDestino = inferDestinoFromCondition(movement.concepto, rule.condition, actions);
    const tablaTrasRegla = actions.tabla_destino ?? classified.tablaDestino;
    if (
      inferredDestino !== undefined &&
      actions.destino === undefined &&
      actions.origen === undefined &&
      tablaTrasRegla === "Gastos"
    ) {
      actions = { ...actions, destino: inferredDestino };
    }
    classified = applyActions(classified, actions);
    lastMatched = rule;
  }

  classified = applyDividendNotasFallback(classified, movement);

  return {
    classification: classified,
    reglaId: lastMatched?.id ?? null,
    reglaNombre: lastMatched?.nombre ?? null,
    reglaPrioridad: lastMatched?.priority ?? -1,
  };
}

function applyDividendNotasFallback(
  classified: Classification,
  movement: PendingMovement,
): Classification {
  if (classified.tablaDestino !== "Ingresos" || classified.notas?.trim()) {
    return classified;
  }
  const origen = classified.origen?.trim();
  if (origen !== "Dividendos" && origen !== "Cashback") {
    return classified;
  }
  const name = movement.metadata.name?.trim();
  if (!name) return classified;
  return { ...classified, notas: name };
}

function ruleMatches(rule: ImportRule, movement: PendingMovement, accountId: string): boolean {
  if (rule.scope === "account" && rule.accountId && rule.accountId !== accountId) {
    return false;
  }

  const condition = rule.condition ?? {};
  if ("importe_positivo" in condition && (movement.importe > 0) !== Boolean(condition.importe_positivo)) {
    return false;
  }
  if ("importe_negativo" in condition && (movement.importe < 0) !== Boolean(condition.importe_negativo)) {
    return false;
  }

  const exact = condition.concepto_exacto;
  if (exact && !matchesConceptoExacto(movement.concepto, exact)) {
    return false;
  }

  const contains = condition.concepto_contiene;
  if (contains && !matchesConceptoContiene(movement.concepto, contains)) {
    return false;
  }

  const regex = condition.concepto_regex;
  if (regex && !matchConceptoRegex(movement.concepto, regex)) {
    return false;
  }

  const txType = condition.tipo ?? condition.type;
  if (txType && movement.metadata.type?.toLowerCase() !== String(txType).toLowerCase()) {
    return false;
  }

  const category = condition.category;
  if (category && movement.metadata.category?.toLowerCase() !== String(category).toLowerCase()) {
    return false;
  }

  const assetClass = condition.asset_class;
  if (assetClass && movement.metadata.asset_class?.toLowerCase() !== String(assetClass).toLowerCase()) {
    return false;
  }

  const mcc = condition.mcc_code;
  if (mcc && movement.metadata.mcc_code !== String(mcc)) {
    return false;
  }

  const symbol = condition.symbol;
  if (symbol && movement.metadata.symbol?.toUpperCase() !== String(symbol).toUpperCase()) {
    return false;
  }

  return true;
}

function applyActions(classified: Classification, actions: RuleActions): Classification {
  const ignorar = actions.ignorar === true ? true : classified.ignorar;
  const tablaDestino = actions.tabla_destino ?? classified.tablaDestino;
  const persona = actions.persona ?? classified.persona;

  let destino = classified.destino;
  let origen = classified.origen;

  if (actions.origen !== undefined) {
    origen = actions.origen || null;
  }
  if (actions.destino !== undefined) {
    const label = actions.destino || null;
    if (tablaDestino === "Ingresos") {
      origen = label;
    } else {
      destino = label;
    }
  }

  const categoria = actions.categoria !== undefined ? actions.categoria : classified.categoria;
  const notas = actions.notas !== undefined ? actions.notas || null : classified.notas;
  const tipo = actions.tipo !== undefined ? actions.tipo : classified.tipo;
  const nombre = actions.nombre !== undefined ? actions.nombre : classified.nombre;
  const entidad = actions.entidad !== undefined ? actions.entidad : classified.entidad;

  let importe = classified.importe;
  if (actions.importe_signo) {
    const sign = actions.importe_signo === "positivo" ? 1 : -1;
    importe = Math.abs(importe) * sign;
  }
  if (actions.invertir_importe) {
    importe = importe * -1;
  }

  return {
    tablaDestino,
    ignorar,
    destino,
    origen,
    notas,
    persona,
    categoria,
    tipo,
    nombre,
    entidad,
    importe,
  };
}

export function parseJsonField<T>(value: unknown): T {
  if (value === null || value === undefined) return {} as T;
  if (typeof value === "object") return value as T;
  if (typeof value === "string") {
    try {
      return JSON.parse(value) as T;
    } catch {
      return {} as T;
    }
  }
  return {} as T;
}

export function normalizePersona(value: unknown, fallback: PersonaValue = "Santi"): PersonaValue {
  if (value === "Santi" || value === "Sandra" || value === "Común") return value;
  return fallback;
}

export function normalizeMetadata(value: unknown): Record<string, string> {
  if (!value || typeof value !== "object") return {};
  const out: Record<string, string> = {};
  for (const [key, raw] of Object.entries(value as Record<string, unknown>)) {
    if (raw !== null && raw !== undefined) out[key] = String(raw);
  }
  return out;
}

File: worker/src/worker/rules/engine.py
"""
Motor de ImportRules — ejecutado por la web (wizard), no por el worker/bot.

El worker solo parsea e inserta pending con Metadatos.
La web carga reglas desde NocoDB y aplica al mostrar/revisar pending.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from decimal import Decimal

from worker.models import ClassifiedMovement, Persona, RawMovement, TablaDestino


@dataclass(frozen=True)
class NocoDbRule:
    """Regla cargada desde NocoDB (estructura provisional)."""

    id: str
    scope: str
    account_id: str | None
    priority: int
    condition: dict
    actions: dict
    active: bool = True


def apply_global_sign_rule(movement: RawMovement, persona: Persona, banco: str, key: str) -> ClassifiedMovement:
    tabla: TablaDestino = "Ingresos" if movement.importe > 0 else "Gastos"
    return ClassifiedMovement(
        fecha=movement.fecha,
        importe=movement.importe,
        concepto=movement.concepto,
        banco=banco,
        persona=persona,
        tabla_destino=tabla,
        categoria=None,
        idempotency_key=key,
        referencia=movement.referencia,
        metadata=dict(movement.metadata),
    )


def apply_rules(
    movement: RawMovement,
    *,
    persona: Persona,
    banco: str,
    account_id: str,
    idempotency_key: str,
    nocodb_rules: list[NocoDbRule] | None = None,
) -> ClassifiedMovement:
    classified = apply_global_sign_rule(movement, persona, banco, idempotency_key)
    if not nocodb_rules:
        return classified

    applicable = [
        rule
        for rule in nocodb_rules
        if rule.active and _rule_matches(rule, movement, account_id)
    ]
    applicable.sort(key=lambda rule: rule.priority)

    for rule in applicable:
        classified = _apply_actions(classified, rule.actions)

    return classified


def apply_rules_to_pending(
    pending: ClassifiedMovement,
    *,
    account_id: str,
    nocodb_rules: list[NocoDbRule],
) -> ClassifiedMovement:
    """Aplica ImportRules a un movimiento ya parseado (desde AutomaticActions + Metadatos)."""
    raw = RawMovement(
        fecha=pending.fecha,
        importe=pending.importe,
        concepto=pending.concepto,
        referencia=pending.referencia,
        metadata=dict(pending.metadata),
    )
    return apply_rules(
        raw,
        persona=pending.persona,
        banco=pending.banco,
        account_id=account_id,
        idempotency_key=pending.idempotency_key,
        nocodb_rules=nocodb_rules,
    )


def _rule_matches(rule: NocoDbRule, movement: RawMovement, account_id: str) -> bool:
    if rule.scope == "account" and rule.account_id not in (None, account_id):
        return False
    condition = rule.condition or {}
    if "importe_positivo" in condition and (movement.importe > 0) != bool(condition["importe_positivo"]):
        return False
    if "importe_negativo" in condition and (movement.importe < 0) != bool(condition["importe_negativo"]):
        return False
    contains = condition.get("concepto_contiene")
    if contains and contains.lower() not in movement.concepto.lower():
        return False
    exact = condition.get("concepto_exacto")
    if exact and movement.concepto.strip().lower() != str(exact).strip().lower():
        return False
    regex = condition.get("concepto_regex")
    if regex and not re.search(str(regex), movement.concepto, re.IGNORECASE):
        return False
    tx_type = condition.get("tipo") or condition.get("type")
    if tx_type and movement.metadata.get("type", "").lower() != str(tx_type).lower():
        return False
    category = condition.get("category")
    if category and movement.metadata.get("category", "").lower() != str(category).lower():
        return False
    asset_class = condition.get("asset_class")
    if asset_class and movement.metadata.get("asset_class", "").lower() != str(asset_class).lower():
        return False
    mcc = condition.get("mcc_code")
    if mcc and movement.metadata.get("mcc_code", "") != str(mcc):
        return False
    symbol = condition.get("symbol")
    if symbol and movement.metadata.get("symbol", "").upper() != str(symbol).upper():
        return False
    return True


def _apply_actions(classified: ClassifiedMovement, actions: dict) -> ClassifiedMovement:
    tabla = actions.get("tabla_destino", classified.tabla_destino)
    persona = actions.get("persona", classified.persona)
    categoria = actions.get("categoria", classified.categoria)
    importe = classified.importe
    if "importe_signo" in actions:
        sign = Decimal("1") if actions["importe_signo"] == "positivo" else Decimal("-1")
        importe = abs(importe) * sign
    if actions.get("invertir_importe"):
        importe = importe * Decimal("-1")
    return ClassifiedMovement(
        fecha=classified.fecha,
        importe=importe,
        concepto=classified.concepto,
        banco=classified.banco,
        persona=persona,
        tabla_destino=tabla,
        categoria=categoria,
        idempotency_key=classified.idempotency_key,
        referencia=classified.referencia,
        metadata=dict(classified.metadata),
    )
